ql/src/java/org/apache/hadoop/hive/ql/queryhistory/QueryHistoryService.java

```diff
@@ -0,0 +1,266 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hive.ql.queryhistory;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Queue;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.LinkedBlockingQueue;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.apache.hadoop.hive.conf.HiveConf;
+import org.apache.hadoop.hive.ql.DriverContext;
+import org.apache.hadoop.hive.ql.QueryState;
+import org.apache.hadoop.hive.ql.ServiceContext;
+import org.apache.hadoop.hive.ql.queryhistory.persist.QueryHistoryPersistor;
+import org.apache.hadoop.hive.ql.queryhistory.schema.QueryHistoryRecord;
+import org.apache.hadoop.hive.ql.queryhistory.schema.QueryHistorySchema;
+import org.apache.hadoop.hive.ql.session.SessionState;
+import org.apache.hive.common.util.ReflectionUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+
+public class QueryHistoryService {
+  private static final Logger LOG = LoggerFactory.getLogger(QueryHistoryService.class);
+
+  private static QueryHistoryService INSTANCE = null;
+
+  private QueryHistoryPersistor queryHistoryPersistor;
+
+  protected final Queue<QueryHistoryRecord> queryHistoryQueue = new LinkedBlockingQueue<>();
+  private int maxBatchSize = 0;
+  private int maxQueueSizeInMemoryBytes = 0;
+
+  private final ExecutorService persistExecutor = Executors.newFixedThreadPool(1,
+      new ThreadFactoryBuilder().setNameFormat("QueryHistoryService persistor thread").build());
+
+  private ServiceContext serviceContext;
+
+  public static QueryHistoryService start(HiveConf inputConf, ServiceContext serviceContext) {
+    HiveConf conf = new HiveConf(inputConf);
+    // we can only hit this codepath if the INSTANCE is already initialized while calling start(...)
+    // for convenient test compatibility, this call is allowed to be idempotent (LOG.warn instead of exception)
+    if (INSTANCE != null) {
+      LOG.warn("There is already a QueryHistoryService instance ({}), returning existing instance", INSTANCE);
+      return INSTANCE;
+    }
+
+    INSTANCE = createServiceInstance(serviceContext, conf);
+    return INSTANCE;
+  }
+
+  private static QueryHistoryService createServiceInstance(ServiceContext serviceContext, HiveConf conf) {
+    LOG.info("Starting QueryHistoryService");
+    QueryHistoryService queryHistoryService = new QueryHistoryService();
+
+    initService(serviceContext, conf, queryHistoryService);
+    return queryHistoryService;
+  }
+
+  @VisibleForTesting
+  static void initService(ServiceContext serviceContext, HiveConf conf, QueryHistoryService queryHistoryService) {
```

Comment:
why use static method? can't we use a singleton constructor with an instance `init` method?

Reply:
will be refactored after creating a proper singleton